### PR TITLE
Unit tests

### PR DESCRIPTION
**Pull request by erikkemperman**
As promised earlier, I have been writing unit tests for the server lib. Unfortunately I am now starting a new project and can probably not spend (a lot of) time on this for the next couple of days / weeks.

So I thought I'd submit what I have, even though it's not entirely complete:

- The coverage for the classes that I have tests for is 100%, but I am still missing QueryService -- which I wasn't really sure how to test to be honest.
- Also not covered are the JSR310 converters, but I was wondering why we don't simply use the ones provided by Spring?

The first commit adds a utility class to assert the contents written to logback. If you like this idea, we could use it in the apps generated by the main generator too.

### CODE REVIEW

**Comment by jdubois** [comment]
Wow, that's nearly 5,000 lines of code, incredible!! I'll have time to review this at the end of next week.
BTW, you should join the core team at some point, would you be interested? See http://www.jhipster.tech/team/

**Comment by erikkemperman** [comment]
@jdubois Well, I had some free time recently :-) Would be honoured to join the team but have to say upfront I can't really promise I will have a lot of time...

**Comment by deepu105** [comment]
@erikkemperman dont worry as long as you can contribute something once in a while it should be fine :)

**Comment by erikkemperman** [comment]
FYI Rebased this PR to resolve merge conflicts

**Comment by erikkemperman** [comment]
@deepu105 Well, having made that slight reservation, like I said it'd be an honour!

**Comment by cbornet** [comment]
We should push a Sonar analysis to sonarcloud.io !

**Comment by erikkemperman** [comment]
@cbornet I haven’t looked at that, would it be complicated to add?

**Comment by erikkemperman** [comment]
@cbornet 
I managed to get something working on my fork (unittest branch):

https://sonarcloud.io/dashboard?id=io.github.jhipster%3Ajhipster

But I guess someone of the jhipster org should generate a token and we add the `mvn sonar:sonar` command to the end of the Travis build? Is that how this works?

**Comment by cbornet** [comment]
Yes we need to set the token in Travis.
84% is already quite good, thanks for your work ! Now let's get the AAA !

**Comment by erikkemperman** [comment]
Someone with privileges will have to generate a token and add to Travis. AAA would be nice, but maybe let's first review if the tests I wrote make sense :-)

**Comment by erikkemperman** [comment]
Added a commit to include some Sonar properties (copied from generator, mostly).

After this, if I understand things properly, a member of the jhipster org can generate a sonar token and make a commit similar to [this one](https://github.com/erikkemperman/jhipster/commit/031b576cb500b8ddfdb2d5c28dc1601067dcf0bc) except of course with different `sonar.organization` and `sonar.login` properties.

**Comment by pascalgrimaud** [comment]
The token need to be in a environment variable in Travis instead of beeing directly in travis file.

We have some work to do with JHipster organization in SonarCloud, which is currently not used.

File: src/main/java/io/github/jhipster/async/ExceptionHandlingAsyncTaskExecutor.java [hunk 1]
```diff
@@ -71,7 +73,7 @@ private Runnable createWrappedRunnable(final Runnable task) {
     }
 
     protected void handle(Exception e) {
-        log.error("Caught async exception", e);
+        log.error(EXCEPTION_MESSAGE, e);
```
Comment (deepu105) [inline src/main/java/io/github/jhipster/async/ExceptionHandlingAsyncTaskExecutor.java]: I'm not a fan of using constants for log messages unless they are reused in many places. I see that you use them for test assertions as well but testing log messages seems a bit too much to me @jdubois  what do you think?

File: src/main/java/io/github/jhipster/config/JHipsterConstants.java [hunk 2]
```diff
@@ -22,23 +22,20 @@
 /**
  * JHipster constants.
  */
-public final class JHipsterConstants {
+public interface JHipsterConstants {
```
Comment (deepu105) [inline src/main/java/io/github/jhipster/config/JHipsterConstants.java]: I thought this was only adding tests but seems you rewrote some stuff as well so I would like @jdubois to have a look as well before merging this

File: src/test/java/io/github/jhipster/CoverageTest.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+import java.util.Set;
+import java.util.TreeSet;
+
+import org.junit.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.reflect.ClassPath;
+
+public class CoverageTest {
```
Comment (deepu105) [inline src/test/java/io/github/jhipster/CoverageTest.java]: can you add some docs here on what the class does

File: src/test/java/io/github/jhipster/CoverageTest.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        if (missing.size() > 0) {
+            StringBuffer buffer = new StringBuffer(64 + 32 * missing.size());
+            buffer.append("Missing " + missing.size() + " classes:" + System.lineSeparator());
+            for (String clazz : missing) {
+                buffer.append("  " + clazz + System.lineSeparator());
+            }
+            log.warn(buffer.toString());
+        }
+        // TODO uncomment:
+        // assertThat(missing).isEmpty();
```
Comment (deepu105) [inline src/test/java/io/github/jhipster/CoverageTest.java]: ??

File: src/main/java/io/github/jhipster/async/ExceptionHandlingAsyncTaskExecutor.java (hunk 1 above)
Comment (erikkemperman) [reply src/main/java/io/github/jhipster/async/ExceptionHandlingAsyncTaskExecutor.java]: Yes, that’s why I made them constants. Could be made package private if you prefer. Or like you say maybe asserting log messages is too much, but for what it’s worth I actually quite like it. Especially for failure messages, logging is an important function and as such ought to be tested.

File: src/main/java/io/github/jhipster/config/JHipsterConstants.java (hunk 2 above)
Comment (erikkemperman) [reply src/main/java/io/github/jhipster/config/JHipsterConstants.java]: Just some little tweaks, like those message constants and things like this which changes a class that’s never instantiated to an interface (helps with coverage).

Other than that just minor refactoring so I could spy or override selectively.

No functional changes I think.

File: src/test/java/io/github/jhipster/CoverageTest.java (hunk 3 above)
Comment (erikkemperman) [reply src/test/java/io/github/jhipster/CoverageTest.java]: I figured the name of the class would be self-explanatory but maybe not :-)

Actually maybe this test is overly pedantic, it traverses the packages and asserts that every regular class has a `...Test` companion.

I’ll await the feedback here, if this test stays I will add a comment about what it does.

File: src/test/java/io/github/jhipster/CoverageTest.java (hunk 4 above)
Comment (erikkemperman) [reply src/test/java/io/github/jhipster/CoverageTest.java]: Once coverage is complete we could uncomment this, it’ll throw a failure when new classes are added but an accompanying test class is missing.

But like I said above, perhaps this is too pedantic.

File: src/main/java/io/github/jhipster/async/ExceptionHandlingAsyncTaskExecutor.java (hunk 1 above)
Comment (erikkemperman) [reply src/main/java/io/github/jhipster/async/ExceptionHandlingAsyncTaskExecutor.java]: Oh and also, the way the log testing works also keeps the console nice and clean during `mvn test`. So expected exceptions during tests, for example, don’t spew out endless traces.

File: src/test/java/io/github/jhipster/CoverageTest.java (hunk 3 above)
Comment (deepu105) [reply src/test/java/io/github/jhipster/CoverageTest.java]: i'm not in favor of doing this ourself, why not us a gradle plugin for that?

Comment (erikkemperman) [reply src/test/java/io/github/jhipster/CoverageTest.java]: Well, Maven. I’m not too familiar with jacoco but if we can make it fail the build if coverage drops significantly, sure.

But don’t read too much into this particular class, it started as more of a todo list for this PR and only at the end did I add the commented assert as an afterthought

No problem to drop this commit.

Comment (deepu105) [reply src/test/java/io/github/jhipster/CoverageTest.java]: Then lets remove this. And I don't want builds to fail if coverage drops not yet atleast.

**Comment by deepu105** [comment]
@jdubois i would like to merge this but would like your eyes on it as well. Can you take a quick look

**Comment by deepu105** [comment]
@erikkemperman what is your full name, twitter handle and email  address so that I can start a formal invitation process to the team
If you dont want to share it here you can send it to d4udts@gmail.com

**Comment by erikkemperman** [comment]
FYI Dropped the commit with the coverage test,, and rebased onto latest master.

@deepu105 I will send some details to the address you mentioned, thanks for following up!

File: pom.xml [hunk 5]
```diff
@@ -42,10 +42,20 @@
...
+        <project.testresult.directory>${project.build.directory}/test-results</project.testresult.directory>
         <maven.build.timestamp.format>yyyyMMddHHmmss</maven.build.timestamp.format>
         <maven.compiler.source>${java.version}</maven.compiler.source>
         <maven.compiler.target>${java.version}</maven.compiler.target>
 
+        <!-- Sonar properties -->
+        <sonar.jacoco.itReportPath>${project.testresult.directory}/coverage/jacoco/jacoco-it.exec</sonar.jacoco.itReportPath>
+        <sonar.jacoco.reportPath>${project.testresult.directory}/coverage/jacoco/jacoco.exec</sonar.jacoco.reportPath>
+        <sonar.java.codeCoveragePlugin>jacoco</sonar.java.codeCoveragePlugin>
+        <sonar.typescript.lcov.reportPaths>${project.testresult.directory}/coverage/report-lcov/lcov.info</sonar.typescript.lcov.reportPaths>
```
Comment (deepu105) [inline pom.xml]: This repo doesn't and will never contain Typescript code so I guess this can be removed

Comment (erikkemperman) [reply pom.xml]: Ah, good catch, thanks! Sloppy/paste from the generator POM. Will amend now!

File: src/main/java/io/github/jhipster/async/ExceptionHandlingAsyncTaskExecutor.java (hunk 1 above)
Comment (jdubois) [reply src/main/java/io/github/jhipster/async/ExceptionHandlingAsyncTaskExecutor.java]: I'd rather also have normal Strings as they are more readable, but that's not a huge issue

**Comment by jdubois** [comment]
There are some formatting issues, etc... But this is so big & good I'm merging this, and I'll improve later.

**Comment by erikkemperman** [comment]
Thanks for review and merge! I tried to follow formatting guidelines, so I’ll keep an eye on changes here so hopefully I’ll get it right next time.

About those strings, is it the constants you don’t like or asserting the contents of logs? Because ai wanted to propose doing something similar in generated code...

One benefit aside from checking the logs is that it would avoid cluttering the console — I’ve been getting some errors in Travis about excessive output.

Thoughts?

**Comment by jdubois** [comment]
Let me first finish to polish everything :-)
But yes I'd rather keep the Strings, it's more readable. Let's keep it like that on the server lib, it's less important, and it allows us to test this a bit.

**Comment by erikkemperman** [comment]
All right ;-)

**Pull request merged**